Optional access lists (#2930)
ethereum/EIPs: eip-2930.md

---
eip: 2930
title: Optional access lists
author: Vitalik Buterin (@vbuterin), Martin Swende (@holiman)

> discussions-to: TBD

Reviewer: Need this before we merge.

Author: Do you think just pointing it to the same place as EIP 2929 discussions is ok? The two EIPs are closely intertwined so talking about them together might be reasonable?

> discussions-to: TBD

Author: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-2930-optional-access-lists/4561
```

status: Draft
type: Standards Track
category: Core
created: 2020-08-29

> requires: EIP 2718; EIP 2929

Reviewer: ```suggestion
requires: 2718, 2929
```

---

## Simple Summary

> Adds a transaction type which contains an access list, a list of addresses and storage keys that the transaction plans to access. Accesses outside the list are possible, but become more expensive. Intended as a mitigation to contract breakage risks introduced by EIP 2929 and simultaneously a stepping stone toward broader use of access lists in other contexts.

Reviewer: ```suggestion
Adds a transaction type which contains an access list, a list of addresses and storage keys that the transaction plans to access. Accesses outside the list are possible, but become more expensive.
```
The latter part of this should move to the motivation section.

## Abstract

> We introduce a new [EIP-2718](https://github.com/ethereum/EIPs/pull/2718) transaction type, with the format `rlp([3, [nonce, gasPrice, gasLimit, to, value, data, access_list, senderV, senderR, senderS]])`.

Author: ```suggestion
We introduce a new [EIP-2718](./eip-2718.md) transaction type, with the format `rlp([3, [nonce, gasPrice, gasLimit, to, value, data, access_list, senderV, senderR, senderS]])`.
```

> We introduce a new [EIP-2718](https://github.com/ethereum/EIPs/pull/2718) transaction type, with the format `rlp([3, [nonce, gasPrice, gasLimit, to, value, data, access_list, senderV, senderR, senderS]])`.

Reviewer: There is currently a debate on whether internal EIP links should be `[EIP-2718](./eip-2718.md)` or `[EIP-2718](https://eips.ethereum.org/EIPS/eip-2718)`.  What you have here isn't either so is definitely not in line with the current "standard".  😄

> The `access_list` specifies a list of addresses and storage keys; these addresses and storage keys are added into the `accessed_addresses` and `accessed_storage_keys` global sets (introduced in https://notes.ethereum.org/@vbuterin/BkrNbeAfD). A gas cost is charged, though at a discount relative to the cost of accessing outside the list.

Author: ```suggestion
The `access_list` specifies a list of addresses and storage keys; these addresses and storage keys are added into the `accessed_addresses` and `accessed_storage_keys` global sets (introduced in [EIP-2929](./eip-2929.md)). A gas cost is charged, though at a discount relative to the cost of accessing outside the list.
```

> The `access_list` specifies a list of addresses and storage keys; these addresses and storage keys are added into the `accessed_addresses` and `accessed_storage_keys` global sets (introduced in https://notes.ethereum.org/@vbuterin/BkrNbeAfD). A gas cost is charged, though at a discount relative to the cost of accessing outside the list.

Reviewer: Recommend against linking out to external documents whenever possible.  The links will likely *eventually* 404, at which point the EIP becomes harder to understand.  Where possible, either inline the content or add it as an asset in `../assets/eip-2930/...`

## Motivation

This EIP serves two functions:

> 1. Mitigates contract breakage risks introduced by https://notes.ethereum.org/@vbuterin/BkrNbeAfD, as transactions could pre-specify and pre-pay for the accounts and storage slots that the transaction plans to access; as a result, in the actual execution, the SLOAD and EXT* opcodes would only cost 100 gas: low enough that it would not only prevent breakage due to that EIP but also "unstuck" any contracts that became stuck due to EIP 1884.

Author: ```suggestion
1. Mitigates contract breakage risks introduced by [EIP-2929](./eip-2929.md), as transactions could pre-specify and pre-pay for the accounts and storage slots that the transaction plans to access; as a result, in the actual execution, the SLOAD and EXT* opcodes would only cost 100 gas: low enough that it would not only prevent breakage due to that EIP but also "unstuck" any contracts that became stuck due to EIP 1884.
```
